File: backend/utils/password_policy.py
"""Shared password strength rules for registration, profile update, and reset."""

from __future__ import annotations

import re

PASSWORD_MIN_LENGTH = 8

PASSWORD_REQUIREMENTS_MESSAGE = (
    "Password must be at least 8 characters and include an uppercase letter, "
    "a lowercase letter, a number, and a special character."
)

_HAS_UPPER = re.compile(r"[A-Z]")
_HAS_LOWER = re.compile(r"[a-z]")
_HAS_DIGIT = re.compile(r"[0-9]")
_HAS_SPECIAL = re.compile(r"[^A-Za-z0-9]")


def is_strong_password(password: str) -> bool:
    if not password or len(password) < PASSWORD_MIN_LENGTH:
        return False
    if not _HAS_UPPER.search(password):
        return False
    if not _HAS_LOWER.search(password):
        return False
    if not _HAS_DIGIT.search(password):
        return False
    if not _HAS_SPECIAL.search(password):
        return False
    return True


def validate_password_strength(password: str) -> str:
    if not is_strong_password(password):
        raise ValueError(PASSWORD_REQUIREMENTS_MESSAGE)
    return password

File: frontend/scripts/utils/passwordPolicy.js
(function (global) {
  'use strict';

  var DEFAULT_MESSAGE =
    'Password must be at least 8 characters and include an uppercase letter, ' +
    'a lowercase letter, a number, and a special character.';

  function isStrong(password) {
    var value = String(password || '');
    if (value.length < 8) return false;
    if (!/[A-Z]/.test(value)) return false;
    if (!/[a-z]/.test(value)) return false;
    if (!/[0-9]/.test(value)) return false;
    if (!/[^A-Za-z0-9]/.test(value)) return false;
    return true;
  }

  function message(i18nKey) {
    if (window.i18n && typeof window.i18n.t === 'function' && i18nKey) {
      var translated = window.i18n.t(i18nKey);
      if (translated && translated !== i18nKey) return translated;
    }
    return DEFAULT_MESSAGE;
  }

  global.PasswordPolicy = {
    isStrong: isStrong,
    message: message,
    defaultMessage: DEFAULT_MESSAGE
  };
})(window);
